_episodes/05-functions.md

```diff
@@ -0,0 +1,484 @@
+>
+> ~~~
+> # revised version
+> import random
+> for i in range(10):
+>
+>     # simulating the mass of a chicken egg
+>     # the (random) mass will be 70 +/- 20 grams
+>     mass = 70 + 20.0 * (2.0 * random.random() - 1.0)
+>
+>     print(mass, get_egg_label(mass))
+>
+> ~~~
+> {: .language-python}
+>
+>
+> 1. Create a function definition for `get_egg_label()` that will work with the revised program above.  Note that the `get_egg_label()` function's return value will be important. Sample output from the above program would be `71.23 large`.
+> 2. A dirty egg might have a mass of more than 90 grams, and a spoiled or broken egg will probably have a mass that's less than 50 grams.  Modify your `get_egg_label()` function to account for these error conditions. Sample output could be `25 too light, probably spoiled`.
+>
+> > ## Solution
+> >
+> > ~~~
+> > def get_egg_label(mass):
+> >     # egg sizing machinery prints a label
+> >     egg_label = "Unlabelled"
+> >     if mass >= 90:
+> >         egg_label = "warning: egg might be dirty"
+> >     elif mass >= 85:
+> >         egg_label = "jumbo"
+> >     elif mass >= 70:
+> >         egg_label = "large"
+> >     elif mass < 70 and mass >= 55:
+> >         egg_label = "medium"
+> >     elif mass < 50:
+> >         egg_label = "too light, probably spoiled"
+> >     else:
+> >         egg_label = "small"
+> >     return egg_label
+> > ~~~
+> > {: .language-python}
+> {: .solution}
+{: .challenge}
+
+> ## Simulating a dynamical system
+>
+> In mathematics, a [dynamical system](https://en.wikipedia.org/wiki/Dynamical_system) is a system
+> in which a function describes the time dependence of a point in a geometrical space. A canonical
+> example of a dynamical system is the [logistic map](https://en.wikipedia.org/wiki/Logistic_map),
+> a growth model that computes a new population density (between  0 and 1) based on the current
+> density. In the model, time takes discrete values 0, 1, 2, ...
+>
+> 1. Define a function called `logistic_map` that takes two inputs: `x`, representing the current
+>    population (at time `t`), and a parameter `r = 1`. This function should return a value 
+>    representing the state of the system (population) at time `t + 1`, using the mapping function:
+>
+>    `f(t+1) = r * f(t) * [1 - f(t)]`
+>
+> 2. Using a `for` or `while` loop, iterate the `logistic_map` function defined in part 1, starting
+>    from an initial population of 0.5, for a period of time `t_final = 10`. Store the intermediate
+>    results in a list so that after the loop terminates you have accumulated a sequence of values
+>    representing the state of the logistic map at times `t = [0,1,...,t_final]` (11 values in total).
+>    Print this list to see the evolution of the population.
+>
+> 3. Encapsulate the logic of your loop into a function called `iterate` that takes the initial
+>    population as its first input, the parameter `t_final` as its second input and the parameter
```

Comment:
Is it obvious that 'input' can be used as a synonym for 'argument?' Since input is more generic, I would tend to think of 'input' as the set of all arguments (so the first argument would be the 'first piece of input' and so on).

I think it's fine to say: "Here, we are using 'input' as a synonym for 'argument.'"

Reply:
I think it's rather clear from the uses of word input in this tutorial, but I will update it as you suggest

Reply:
Now I'm thinking that the word 'parameter' may cause confusion:

* There are 'parameters' in a function's definition in the context of this lesson (in computer programming);
* but, in this sentence, it seems that 'parameter' is used as in *mathematics*, referring to the `r` parameter in this equation `f(t+1) = r * f(t) * [1 - f(t)]`, for example.

For clarity, I think I would simplify the sentence to something like "... write a function ... where the second parameter is `t_final` ..."